### PR TITLE
timer: cortex-m systick: add idle timer

### PR DESCRIPTION
**Pull request by niedzwiecki-dawid**
Some chips, that use Cortex-M SysTick as the system timer, disable a clock in a low power mode, that is the input for the SysTick e.g. STM32Fx family.

It blocks enabling power management for these chips. The wake-up function doesn't work and the time measurement is lost.

Add an additional IDLE timer that handles these functionality when the system is about to enter IDLE. It has to wake up the chip and update the cycle counter by time not measured by the SysTick. The IDLE timer has to support counter API (setting alarm and reading current value).

Related:
#61631 
#60558
#19755

### CODE REVIEW

**Review by erwango — COMMENTED** [review]
Thanks for submitting this. I have a quite similar need for a different use case:

In other STM32 series which have a low power timer which helps counting tick in lp modes (`stop ` modes in STM32 nomenclature), we would need the same functionality to enter deeper sleep states  (`standby` mode) where even LP timer would lose its clock and only RTC could be used (common point with your use case).

My understanding is that to implement my use case, one option would be to reproduce the nearly* same piece of code in `drivers/timer/stm32_lptim_timer.c`. Then maybe others would need the same implementation in their own timer driver.

So here is my (not blocking) suggestion: 
Since this piece of code is destined to be duplicated in multiple timer drivers, I wonder if it could be implemented as a "generic timer back up service", rather than a specific cortex_m_systick one.

*Note: in my case, I would not need an extra logic:
Using lptimer, I would be able to enter some of the sleep states and only the deepest one would need to trigger use of counter/RTC wakeup. So the `idle` parameter will not be sufficient and I would need way to distinguish next state: stop or standby.
I guess this extra functionality can be added later on in a "generic timer back up service".

**Comment by semihalf-niedzwiecki-dawid** [comment]
Thanks for the comment.

Very interesting case. However, I've tried to extract common code of the "back up service" for all timers and my conclusion is that it is not a good idea. 2 main points: the code uses a local variable, and it is important to read values of the two timers as soon as possible (no calculation between). 

The one thing would make sense, I think I can change the config and chosen string to `CORTEX_IDLE_TIMER`, to use it by all drivers.

Please let me know what you think.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Do you know who would be able to review this PR?

**Comment by erwango** [comment]
> 2 main points: the code uses a local variable, and it is important to read values of the two timers as soon as possible (no calculation between).

What about setting variable scope to global?

**Comment by erwango** [comment]
> Do you know who would be able to review this PR?

@anangl and @andyross

**Comment by erwango** [comment]
and .. @teburd :) !

**Comment by semihalf-niedzwiecki-dawid** [comment]
>Using lptimer, I would be able to enter some of the sleep states and only the deepest one would need to trigger use of counter/RTC wakeup. So the idle parameter will not be sufficient and I would need way to distinguish next state: stop or standby.

I've been thinking about your case. IMO, there is a simply way to handle that. You can compare the measurement of IDLE and system timers. If a difference is bigger than ~10%/1-2 IDLE timer ticks, it means the system timer was stopped and needs correction. Does it make sense?

**Comment by erwango** [comment]
> > Using lptimer, I would be able to enter some of the sleep states and only the deepest one would need to trigger use of counter/RTC wakeup. So the idle parameter will not be sufficient and I would need way to distinguish next state: stop or standby.
> 
> I've been thinking about your case. IMO, there is a simply way to handle that. You can compare the measurement of IDLE and system timers. If a difference is bigger than ~10%/1-2 IDLE timer ticks, it means the system timer was stopped and needs correction. Does it make sense?

Actually my point is that the timer to use depends on the sleep duration (defined by low power state residency). Since standby mode (and use of RTC counter) would require longer time to exit (and potentially also enter), I would go in standby mode only if time allows and for shorter timeout periods, I would use STOP Mode / LPTIM. This knowledge is available in `sys_clock_set_timeout` caller.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Make sense

File: drivers/timer/cortex_m_systick.c
Comment (erwango) [inline drivers/timer/cortex_m_systick.c:389]: Can't you call `sys_clock_announce()` at that point ?

Comment (semihalf-niedzwiecki-dawid) [reply drivers/timer/cortex_m_systick.c:389]: It’s a very good point. I’ve been considering two solutions. I feel like the current solution is safer, the announcement is in one place, code is not repeated, cycles to announce are aligned to tick boundary. However it introduces an additional delay. According to my measurement, delay between `sys_clock_idle_exit` and `sys_clock_isr` (after `sys_clock_set_timeout(0, false)`) is 1-2 RTC ticks (62,5us - 125us), so quite significant. I was surprised about that, alignment to the tick boundary(100us in my system) may be the reason. I’m ready to move the announcement here, if you don’t see any problems.

Comment (semihalf-niedzwiecki-dawid) [reply drivers/timer/cortex_m_systick.c]: I've changed the way of announcement. I haven't encountered any issues with it.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Friendly ping :-) Please review if you can, this change is very important for my work.

**Comment by teburd** [comment]
This is a pretty significant change to the cortex-m timer, and that's a widely used critical piece of software in Zephyr.

@npitre is honestly probably the best person to review this

**Comment by semihalf-niedzwiecki-dawid** [comment]
> This is a pretty significant change to the cortex-m timer, and that's a widely used critical piece of software in Zephyr.
> 
> I get the idea here, though I'm unsure about this idea of the timer depending on the counter API and driver.
> 
> @npitre is honestly probably the best person to review this

That's true :-) However, the change impacts only users that add the idle timer. Each part of code I've added is inside ifdef `#ifdef CONFIG_CORTEX_M_SYSTICK_IDLE_TIMER`.  Of course it doesn't mean it can be buggy.

**Comment by npitre** [comment]
What does the `idle_lock` protect against?

**Comment by semihalf-niedzwiecki-dawid** [comment]
> What does the `idle_lock` protect against?

When we enter the `sys_clock_idle_exit` function, the SysTick is already running. The LOAD register hasn’t been changed while setting the timeout, so we can not be sure about its value. That means the SysTick IRQ can occur any time. `cycle_count` is updated in the `sys_clock_idle_exit` so we can not be interrupted. The second important thing is that these two measurements 
```
		counter_get_value(idle_timer, &idle_timer_post);
		systick_diff = cycle_count + elapsed() - cycle_pre_idle;
```
have to happen one after another, without any interruption.

**Comment by npitre** [comment]
> > What does the `idle_lock` protect against?
> 
> When we enter the `sys_clock_idle_exit` function, the SysTick is already
> running. The LOAD register hasn’t been changed while setting the timeout,
> so we can not be sure about its value. That means the SysTick IRQ can
> occur any time. `cycle_count` is updated in the `sys_clock_idle_exit` so
> we can not be interrupted.

Still, my question remains. Unless you fear concurrent reentrancy into
`sys_clock_idle_exit()`, the `idle_lock` gives you no protection other than
simply disabling interrupts on the local CPU. I doubt there are many SMP
systems using the systick timer out there but still. It looks to me that
you'd better be using the existing `lock` instead.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Thanks for the responses. Disabling interrupts was what I needed, but after making sure the `sys_clock_idle_exit` is called with [the interrupts disabled](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/core/cortex_m/isr_wrapper.S#L54), the lock is not needed anymore.

Anyway, the existing `lock` couldn't be used, because the `sys_clock_announce` function is called, which calls the  `sys_clock_set_timeout` function that uses the `lock`.

**Review by npitre — APPROVED** [review]
I don't fully understand it all and I don't have access to affected platforms, but those changes are all confined inside CONFIG_CORTEX_M_SYSTICK_IDLE_TIMER so that is safe to those who don't need it.

**Comment by semihalf-niedzwiecki-dawid** [comment]
> I don't fully understand it all and I don't have access to affected platforms, but those changes are all confined inside CONFIG_CORTEX_M_SYSTICK_IDLE_TIMER so that is safe to those who don't need it.

Thanks :-) Is there anything I should explain more in details?

**Comment by semihalf-niedzwiecki-dawid** [comment]
Hello :-) do you know anyone else who could review the PR and approve it? Or I should do something more.

**Comment by teburd** [comment]
> Hello :-) do you know anyone else who could review the PR and approve it? Or I should do something more.

This needs maintainer approval (@andyross)

**Comment by niedzwiecki-dawid** [comment]
@andyross can I somehow speed up review? :-)

File: drivers/timer/Kconfig.cortex_m_systick
Comment (andyross) [inline drivers/timer/Kconfig.cortex_m_systick:55]: Would be good to document the required interface for this somewhere.  "The idle timer must be a counter device chosen in device tree via...", etc...

**Review by andyross — APPROVED** [review]
Came here on a ping.  This all seems reasonable to me, a few notes, but the logic seems fine.

File: drivers/timer/cortex_m_systick.c
Comment (niedzwiecki-dawid) [reply drivers/timer/cortex_m_systick.c:98]: Good point, but I would like to keep the kconfig with the help for additional documentation. If you want, I can change the config to non user-configurable one as a follow-up PR.  Does it make sense?

**Pull request merged**
